ocp_resources/utils.py

```diff
@@ -25,76 +30,219 @@ class TimeoutSampler:
+
+        if exceptions_dict:
+            output.update(exceptions_dict)
+
+        elif exceptions is None:
+            output[Exception] = []
+
+        else:
+            if not isinstance(exceptions, tuple):
+                exceptions = (exceptions,)
+
+            for exp in exceptions:
+                if exceptions_msg:
+                    output[exp] = [exceptions_msg]
+                else:
+                    output[exp] = []
+
+        return output
+
     @property
     def _func_log(self):
+        """
+        Returns:
+            string: `func` information to include in log message
+        """
         return f"Function: {self.func} Kwargs: {self.func_kwargs}"
 
     def __iter__(self):
-        last_exp = None
+        """
+        Iterator
+
+        Yields:
+            any: Return value from `func`
+        """
         timeout_watch = TimeoutWatch(timeout=self.wait_timeout)
         if self.print_log:
             LOGGER.info(
                 f"Waiting for {self.wait_timeout} seconds, retry every {self.sleep} seconds"
             )
 
+        last_exp = None
         while timeout_watch.remaining_time() > 0:
             try:
                 self.elapsed_time = self.wait_timeout - timeout_watch.remaining_time()
                 yield self.func(**self.func_kwargs)
                 self.elapsed_time = None
+
                 time.sleep(self.sleep)
 
-            except self.exception as exp:
+            except self._exceptions as exp:
                 last_exp = exp
-                self.elapsed_time = None
+                last_exp_log = self._get_exception_log(exp=last_exp)
+                if self._is_raisable_exception(exp=last_exp):
+                    LOGGER.error(last_exp_log)
+                    raise exp
 
+                self.elapsed_time = None
                 time.sleep(self.sleep)
 
             finally:
                 if self.elapsed_time and self.print_log:
                     LOGGER.info(f"Elapsed time: {self.elapsed_time}")
 
-        raise TimeoutExpiredError(self._process_execution(exp=last_exp))
+        raise TimeoutExpiredError(self._get_exception_log(exp=last_exp))
 
-    def _process_execution(self, exp=None):
-        exp_name = exp.__class__.__name__ if exp else "N/A"
-        last_exception_log = f"Last exception: {exp_name}: {exp}"
+    @staticmethod
+    def _is_exception_matched(exp, exception_messages):
+        """
+        Args:
+            exp (Exception): Exception object raised by `func`
+            exception_messages (list): Either an empty list allowing all text,
+                                       or a list of allowed strings to match against the exception text.
 
-        log = "{timeout}\n{func_log}\n{last_exception_log}".format(
-            timeout=self.wait_timeout,
-            func_log=self._func_log,
-            last_exception_log=last_exception_log,
-        )
+        Returns:
+            bool: True if exception text is allowed, False otherwise
```

Comment:
True if exception text is allowed or not exception text given (in this case the exception as allowed with any massage)